README.md

```diff
@@ -3,47 +3,90 @@
 The vm-fdt crate provides the ability to write Flattened Devicetree blobs as defined in the
 [Devicetree specification](https://devicetree-specification.readthedocs.io/en/stable/flattened-format.html).
 
+In projects such as
+[Crosvm](https://chromium.googlesource.com/chromiumos/platform/crosvm/) and
+[Firecracker](https://github.com/firecracker-microvm/firecracker) the device
+tree is used to specify the virtual machine topology (memory, vcpu, caches,
+interrupts, and others) when booting the OS.
+
 ## Design
 
+In vm-fdt we define and work with the following primitives:
+- FDT writer (`FdtWriter`),
+- nodes (`FdtWriterNode`), and
+- properties (defined as key-value pairs).
+
+An FDT has one or multiple nodes, and each node can optionally have
+properties.
+
 The [`FdtWriter` structure](src/writer.rs) provides an interface suitable
-for dynamically generating a Devicetree blob at runtime.
+for dynamically generating a Devicetree blob at runtime. The supported
+operations are:
 
-## Usage
+1. **Creating a node**. This is done by calling `begin_node` on the `FdtWriter`.
+   The call returns an object of type `FdtWriterNode` on which we can set
+   properties (using the `property_*` functions as defined below), or add
+   child nodes (via nested calls to `begin_node`). Properties for the node
+   can only be added *before* creating children nodes (as defined in the
+   specification). Each node must end with a call to the end_node function.
+2. **Create properties**. Properties are key-value pairs, where the key is a
+   string, and the value is a raw bytes array. To make it easier to use, such
+   that the caller of the FDT interface does not need to create a raw byte
+   array, we define wrappers over common property types as follows:
+  * `property_null` (an empty property)
+  * `property_string`
+  * `property_string_list`
+  * `property_u32`
+  * `property_64`
+  * `property_array_u32`
+  * `property_array_u64`
+  * `property` (raw byte array)
 
-TODO: This section describes how the crate is used.
 
-Some questions that might help in writing this section:
-- What traits do users need to implement?
-- Does the crate have any default/optional features? What is each feature
-  doing?
-- Is this crate used by other rust-vmm components? If yes, how?
+## Usage
 
-This crate defines a development feature: `long_running_test`. This feature
-SHOULD NOT be used in production as it might enable functionality that is safe
-only for development use cases.
+The following code is creating an FDT blob with a root node that has 3
+properties (“compatible”, “#address-cells”, and “#size-cells”), and contains
```

Comment:
This text has some fancy quotes in it (“” vs "") - not sure if that matters (e.g. if somebody tries to copy and paste the strings into code.

Reply:
Fixed.